### PR TITLE
Core Bot Sample working -- LuisRecognizer testing pending

### PR DESCRIPTION
**Pull request by axelsrz**
This PR include several small fixes in the libraries (mostly dialogs and core) necessary for the CoreBot sample. The sample omits one step in the BookingDialog given that there is no ChoiceRecognizer ported yet for Python. The LuisRecognizer will be the next part to test.

*Its important to note that everything in the datatypes_timex_expression folder was imported directly from https://github.com/Microsoft/Recognizers-Text/tree/johtaylo/python-timex/Python/libraries/datatypes-timex-expression for the Timex type used to recognize dates

### CODE REVIEW

File: libraries/botbuilder-core/botbuilder/core/bot_state.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ def get(self, turn_context: TurnContext) -> Dict[str, object]:
         if force or (cached_state != None and cached_state.is_changed == True):
             storage_key = self.get_storage_key(turn_context)
             changes : Dict[str, object] = { storage_key: cached_state.state }
+            print(changes)
```
Comment (daveta) [inline libraries/botbuilder-core/botbuilder/core/bot_state.py]: >print(changes) [](start = 12, length = 14)

May want to remove..

File: samples/Core-Bot/datatypes_timex_expression/time.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+class Time:
```
Comment (daveta) [inline samples/Core-Bot/datatypes_timex_expression/time.py]: Nit: Might want to put copyright headers in all stuff for datatypes_timex_expression

File: samples/Core-Bot/datatypes_timex_expression/timex.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            if not hasattr(self, '__time'):
+                setattr(self, '__time', Time(0, 0, value))
+            else:
+                getattr(self, '__time').second = value
+        else:
+            if hasattr(self, '__time'):
+                delattr(self, '__time')
+
+    def clone(self):
+        result = Timex()
```
Comment (daveta) [inline samples/Core-Bot/datatypes_timex_expression/timex.py]: >result = Timex() [](start = 8, length = 16)

Random Q: Would be possible to use copy.copy here?

File: samples/Core-Bot/datatypes_timex_expression/date_range.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
+class DateRange:
+
+    def __init__(self, start, end):
```
Comment (daveta) [inline samples/Core-Bot/datatypes_timex_expression/date_range.py]: >start [](start = 23, length = 5)

Random q: Are these assumed to be datetimes? 
(We've been sort of following [https://www.python.org/dev/peps/pep-0526/](https://www.python.org/dev/peps/pep-0526/).  Highly possible this is not "pythonic" enough, but we should probably be consistent.)

File: samples/Core-Bot/dialogs/booking_dialog.py [hunk 5]
```diff
@@ -35,6 +42,7 @@ def __init__(self, dialog_id: str = None):
     async def origin_step(self, step_context: WaterfallStepContext) -> DialogTurnResult: 
         booking_details = step_context.options
 
+        print('origin_step')
```
Comment (daveta) [inline samples/Core-Bot/dialogs/booking_dialog.py]: >print [](start = 8, length = 5)

Nit: May want to remove.

File: samples/Core-Bot/dialogs/booking_dialog.py [hunk 6]
```diff
@@ -4,26 +4,33 @@
 from .cancel_and_help_dialog import CancelAndHelpDialog
 from .date_resolver_dialog import DateResolverDialog
 
+from datatypes_timex_expression.timex import Timex
+
 class BookingDialog(CancelAndHelpDialog):
 
     def __init__(self, dialog_id: str = None):
         super(BookingDialog, self).__init__(dialog_id or BookingDialog.__name__)
```
Comment (daveta) [inline samples/Core-Bot/dialogs/booking_dialog.py]: >dialog_id or BookingDialog.__name__ [](start = 44, length = 35)

Really like this!

File: samples/Core-Bot/dialogs/date_resolver_dialog.py [hunk 7]
```diff
@@ -3,21 +3,23 @@
...
+from datatypes_timex_expression.timex import Timex
+
 class DateResolverDialog(CancelAndHelpDialog):
 
     def __init__(self, dialog_id: str = None):
         super(DateResolverDialog, self).__init__(dialog_id or DateResolverDialog.__name__)
 
         self.add_dialog(DateTimePrompt(DateTimePrompt.__name__, DateResolverDialog.datetime_prompt_validator))
-        self.add_dialog(WaterfallDialog(WaterfallDialog.__name__, [
+        self.add_dialog(WaterfallDialog(WaterfallDialog.__name__ + '2', [
```
Comment (daveta) [inline samples/Core-Bot/dialogs/date_resolver_dialog.py]: >+ '2' [](start = 65, length = 5)

Curious why the 2..

File: samples/Core-Bot/dialogs/booking_dialog.py [hunk 8]
```diff
@@ -4,26 +4,33 @@
...
+            #self.confirm_step,
+            self.final_step
         ]))
 
-        self.initial_dialog_id(WaterfallDialog.__name__)
+        self.initial_dialog_id = WaterfallDialog.__name__
     
     """
     If a destination city has not been provided, prompt for one.
     """
```
Comment (daveta) [inline samples/Core-Bot/dialogs/booking_dialog.py]: Complete nit: We were attempting this style of docstring:
[https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
But then autorest appears to be adopting reStructuredText, so we started following that to be consistent..

**Comment by daveta** [comment]
Awesome job @axelsrz !!   Really found some big issues with this.  Thank you!

**Comment by congysu** [comment]
>     ):

optional: I missed a type hint ` -> Activity:`.  Thanks.

---
Refers to: libraries/botbuilder-dialogs/botbuilder/dialogs/choices/choice_factory.py:115 in 0fba8fc. [](commit_id = 0fba8fc101941288e7ab56bfcc9553b58e41da78, deletion_comment = False)

**Pull request merged**